src/main/java/dev/j3fftw/litexpansion/machine/MetalForge.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    @Nonnull
+    @Override
+    public List<ItemStack> getDisplayRecipes() {
+        final List<ItemStack> items = new ArrayList<>();
+
+        for (int i = 0; i < recipes.size() - 1; i += 2) {
+            items.add(recipes.get(i)[0]);
+            items.add(recipes.get(i + 1)[0]);
+        }
+
+        return items;
+    }
+
+    protected Inventory createVirtualInventory (Inventory inv){
+        Inventory fakeInv = Bukkit.createInventory(null, 9, "Fake Inventory");
+
+        for (int j = 0; j < inv.getContents().length; j++) {
+            ItemStack stack = inv.getContents()[j] != null && inv.getContents()[j].getAmount() > 1 ?
+                new CustomItem(inv.getContents()[j], inv.getContents()[j].getAmount() - 1) : null;
+            fakeInv.setItem(j, stack);
+        }
+
+        return fakeInv;
+    }
+
+    @Override
+    public void onInteract(Player p, Block b) {
+        Block dispBlock = b.getRelative(BlockFace.DOWN);
+        Dispenser disp = (Dispenser) dispBlock.getState();
+        Inventory inv = disp.getInventory();
+        final List<ItemStack[]> inputs = RecipeType.getRecipeInputList(this);
+
+        for (ItemStack[] input : inputs) {
+            if (canCraft(inv, input)) {
+                final ItemStack output = RecipeType.getRecipeOutputList(this, input).clone();
+
+                if (Slimefun.hasUnlocked(p, output, true)) {
+                    final Inventory fakeInv = createVirtualInventory(inv);
+                    final Inventory outputInv = findOutputInventory(output, dispBlock, inv, fakeInv);
+
+                    if (outputInv != null) {
+                        craft(p, b, inv, input, output, outputInv);
+                    } else
+                        SlimefunPlugin.getLocalization().sendMessage(p, "machines.full-inventory", true);
+                }
+                return;
+            }
+        }
+
+        SlimefunPlugin.getLocalization().sendMessage(p, "machines.unknown-material", true);
+    }
+
+    private void craft(Player p, Block b, Inventory inv, ItemStack[] recipe, ItemStack output, Inventory outputInv) {
+        for (ItemStack removing : recipe) {
+            if (removing != null) {
+                InvUtils.removeItem(inv, removing.getAmount(), true, stack ->
+                    SlimefunUtils.isItemSimilar(stack, removing, true));
+            }
+        }
+
+        outputInv.addItem(output);
+        p.getWorld().playSound(p.getLocation(), Sound.BLOCK_ANVIL_BREAK, 1, 1);
+
+        Block diamondBlock = b.getRelative(BlockFace.DOWN, 2);
+        diamondBlock.setType(Material.AIR);
+    }
+
+    private boolean canCraft(Inventory inv, ItemStack[] recipe) {
+        for (int j = 0; j < inv.getContents().length; j++) {
+            if (!SlimefunUtils.isItemSimilar(inv.getContents()[j], recipe[j], true)) {
+                if (!SlimefunUtils.isItemSimilar(inv.getContents()[j], recipe[j], false)) {
+                    return false;
+                }
+            }
```

Comment:
```suggestion
            if (!SlimefunItem.getById(inv.getContents()[j]).getID().equals(SlimefunItem.getById(recipe[j]).getID())) {
                return false;
            }
```
Also, a null check would be good before